### PR TITLE
Single precision interpolation

### LINKED ISSUE #1357: MPI with SparseFunction on the mpi domain boundary
There seem to be an issue when the source is on or near the MPI halo in some cases.

MFE  to be made.

### PR DESCRIPTION
**Pull request by mloubout**
Fixes interpolation using the wrong points with MPI (or without MPI and no BCs) due to rounding errors around zero in single precision.

- Added test for it

Fixes #1357 

Should fix #1251

### CODE REVIEW

File: devito/types/sparse.py [hunk 1]
```diff
@@ -535,6 +536,27 @@ def _point_symbols(self):
...
+        """
+        Symbols map between grid origin and sparse point position
+        along each SpaceDimension.
+
+        Notes
+        -----
+        The expression `(coord - origin)/spacing` could also be computed in the
+        mathematically equivalent expanded form `coord/spacing -
+        origin/spacing`. This particular form is problematic when a sparse
+        point is in proximity of the grid origin, since due to a larger machine
```
Comment (rhodrin) [inline devito/types/sparse.py:550]: "...is in close proximity..."

**Comment by rhodrin** [comment]
Looks good thanks. Are there any more tests we should add?

Need to think about a couple of issues (e.g. how this will work with the additional shift with `Function`'s on `SubDomain`'s) but can do that at a later date (when the books sprint is over).

File: devito/types/sparse.py [hunk 2]
```diff
@@ -535,6 +536,27 @@ def _point_symbols(self):
         return tuple(Scalar(name='p%s' % d, dtype=self.dtype)
                      for d in self.grid.dimensions)
 
+    @cached_property
+    def _position_map(self):
+        """
+        Symbols map between grid origin and sparse point position
```
Comment (rhodrin) [inline devito/types/sparse.py:542]: "Symbolic map between the grid origin and sparse point position(s)..."?

Comment (mloubout) [reply devito/types/sparse.py:542]: I prefer what I had, this sounds like it maps `{ origin : coordinate}` which it's something else.

File: devito/types/sparse.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/types/sparse.py]: done

File: devito/types/sparse.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/types/sparse.py]: reverted

**Comment by FabioLuporini** [comment]
> Looks good thanks. Are there any more tests we should add?

not sure tbh?

**Pull request merged**